crates/bevy_math/src/primitives/dim2.rs

```diff
@@ -0,0 +1,181 @@
+    pub fn from_normalized(value: Vec2) -> Self {
+        debug_assert!(value.is_normalized());
+        Self(value)
+    }
+}
+
+impl std::ops::Deref for Direction2d {
+    type Target = Vec2;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// A circle primitive
+#[derive(Clone, Copy, Debug)]
+pub struct Circle {
+    /// The radius of the circle
+    pub radius: f32,
+}
+impl Primitive2d for Circle {}
+
+/// An unbounded plane in 2D space. It forms a separating surface trough the origin,
+/// stretching infinitely far
+#[derive(Clone, Copy, Debug)]
+pub struct Plane2d {
+    /// The normal of the plane, the plane will be placed perpendicular to this direction
+    pub normal: Direction2d,
+}
+impl Primitive2d for Plane2d {}
+
+/// An infinite line along a direction in 2D space.
+///
+/// For a finite line: [`Segment2d`]
+#[derive(Clone, Copy, Debug)]
+pub struct Line2d {
+    /// The direction of the line, the line extends infinitely in both the positive
+    /// and negative of this direction
+    pub direction: Direction2d,
+}
+impl Primitive2d for Line2d {}
+
+/// A segment of a line along a direction in 2D space.
+#[doc(alias = "LineSegment2d")]
+#[derive(Clone, Debug)]
+pub struct Segment2d {
+    /// The direction of the line
+    pub direction: Direction2d,
+    /// Half the length of the line segment, the segment extends by this amount in both
+    /// the positive and negative direction
+    pub half_length: f32,
+}
+impl Primitive2d for Segment2d {}
+
+impl Segment2d {
+    /// Create a line segment from a direction and full length of the line
+    pub fn new(direction: Direction2d, length: f32) -> Self {
+        Self {
+            direction,
+            half_length: length / 2.,
+        }
+    }
+
+    /// Get a line segment and translation from two points at each end of a line segment
+    ///
+    /// Panics if point1 == point2
+    pub fn from_points(point1: Vec2, point2: Vec2) -> (Self, Vec2) {
+        let diff = point2 - point1;
+        let length = diff.length();
+        (
+            Self::new(Direction2d::from_normalized(diff / length), length),
+            (point1 + point2) / 2.,
+        )
+    }
+
+    /// Get the position of the first point on the line segment
+    pub fn point1(&self) -> Vec2 {
+        *self.direction * -self.half_length
+    }
+
+    /// Get the position of the second point on the line segment
+    pub fn point2(&self) -> Vec2 {
+        *self.direction * self.half_length
+    }
+}
+
+/// A series of connected line segments in 2D space.
+///
+/// For a version without generics: [`BoxedPolyline2d`]
+#[derive(Clone, Debug)]
+pub struct Polyline2d<const N: usize> {
+    /// The vertices of the polyline
+    pub vertices: [Vec2; N],
+}
+impl<const N: usize> Primitive2d for Polyline2d<N> {}
+
+/// A series of connected line segments in 2D space.
+///
+/// For a version without alloc: [`Polyline2d`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolyline2d {
```

Comment:
Since we are talking about shapes, doesn't the name `Boxed` will confound the users? I mean, doesn't this sound like a "line 2d in space witch forms a box"?

Reply:
I'm not sure what else it'd be called, and the idea of having a separate type for "a polyline that forms a box" would be weird and niche enough that I don't think people would actually expect this to be that. "Boxed" is also a relatively common term in Rust (the docs use it). The docs could explicitly say that the vertices are heap-allocated with a `Box<[Vec2]>` to make it clearer though.